src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Unix.cs

```diff
@@ -276,12 +279,149 @@ internal unsafe SocketError DoOperationSend(SafeCloseSocket handle)
+            Debug.Assert(_sendPacketsElements != null);
+            SendPacketsElement[] elements = (SendPacketsElement[])_sendPacketsElements.Clone();
+            FileStream[] files = new FileStream[elements.Length];
+
+            // Open all files synchronously ahead of time so that any exceptions are propagated
+            // to the caller, to match Windows behavior.
+            try
+            {
+                for (int i = 0; i < elements.Length; i++)
+                {
+                    string path = elements[i]?.FilePath;
+                    if (path != null)
+                    {
+                        files[i] = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, 0x1000, useAsync: true);
+                    }
+                }
+            }
+            catch (Exception exc)
+            {
+                // Clean up any files that were already opened.
+                foreach (FileStream s in files)
+                {
+                    s?.Dispose();
+                }
+
+                // Windows differentiates the directory not being found from the file not being found.
+                // Approximate this by checking to see if the directory exists; this is only best-effort,
+                // as there are various things that could affect this, e.g. directory creation racing with
+                // this check, but it's good enough for most situations.
+                if (exc is FileNotFoundException fnfe)
+                {
+                    string dirname = Path.GetDirectoryName(fnfe.FileName);
+                    if (!string.IsNullOrEmpty(dirname) && !Directory.Exists(dirname))
+                    {
+                        throw new DirectoryNotFoundException(fnfe.Message);
+                    }
+                }
+
+                // Otherwise propagate the original error.
+                throw;
+            }
+
+            SendPacketsAsync(socket, elements, files);
+            return SocketError.IOPending;
+        }
+
+        private async void SendPacketsAsync(Socket socket, SendPacketsElement[] elements, FileStream[] files) // async void, as completion is communicated via other means
+        {
+            try
+            {
+                int bytesTransferred = 0;
+                Debug.Assert(elements.Length == files.Length);
+                for (int i = 0; i < elements.Length; i++)
+                {
+                    SendPacketsElement e = elements[i];
+                    if (e != null)
+                    {
+                        if (e.FilePath == null)
+                        {
+                            bytesTransferred += await socket.SendAsync(new ArraySegment<byte>(e.Buffer, e.Offset, e.Count), SocketFlags.None).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            FileStream fs = files[i];
+                            Debug.Assert(fs != null);
```

Comment:
Nit: is it worth the bother of asserting something's not null when it will throw NRE immediatley after.